bravado_core/operation.py

```
@@ -5,10 +5,29 @@
 from bravado_core.exception import SwaggerSchemaError
 from bravado_core.param import Param
 from bravado_core.security_requirement import SecurityRequirement
+from bravado_core.util import cached_property
 
 log = logging.getLogger(__name__)
 
 
+def _sanitize_operation_id(operation_id, http_method, path_name):
+    if operation_id is None:
+        # build based on the http method and request path
+        operation_id = http_method + '_' + path_name
+
+    for regex, replacement in (
+            ('[^A-Za-z0-9_]', '_'),  # valid chars for method names
+            ('_+', '_'),  # collapse consecutive _'s
+            ('^_|_$', '')):  # trim leading/trailing _'s
+        operation_id = re.compile(regex).sub(replacement, operation_id)
+
+    # Handle crazy corner cases where someone explictily sets operation
+    # id a value that gets sanitized down to an empty string
+    if len(operation_id) == 0:
+        operation_id = _sanitize_operation_id(None, http_method, path_name)
```

Note that there is a (theoretical?) corner case where this leads to an infinite loop. If operation_id is already None we should throw an Exception or return a different result, but not recurse again.

---

``operation_id`` is ``None`` if it is not specified in the swagger specs.
This method will replace ``operation_id`` with ``http_method + '_' + path_name`` (ie. ``test_returns_generated_operation_id_when_missing_from_operation_spec`` test) and the new operation ``id`` cannot be ``None`` or sanitised to an empty string.

---

Well, if http_method is empty for any reason and path_name only contains '_' or invalid chars then this will result in an infinite loop I think. I agree that will probably never happen (outside of bugs), but let's make sure. :)

---

This case is only possible in case of invalid specs.
We could be more conservative in this case by asserting that http_method is one of ``get``, ``put``, ``post``, ``delete``, ``options``, ``head``, ``patch`` (as specified by [pathItemObject](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md#pathItemObject)).

---

Sounds good! I'm also happy with just throwing an exception in that case if it makes less work (simply because that is so much more obvious to debug than an infinite loop / hang). If we save the operation_id as passed in then we can check at this point if it was None originally - in that case there is no need for the recursive call, as it will not yield different results. We should just abort.

---

Refactored a bit how it performs the sanitisation operations so we can prevent recursive calls :)